### PR TITLE
CIP-0052? | Cardano audit best practice guidelines

### PR DESCRIPTION
**Pull request by simonjohnthompson**
These guidelines are the result of a process of discussion between IOG staff, and members of the audit and academic communities, over a series of online meetings in February and March 2022. Audit organisations involved include Tweag, WellTyped, Certik, Runtime Verification, BT Block, MLabs, Quviq and Hachi/Meld, all of which supported the guidelines outlined here. 

The CIP has been in the forum for one week now. It would be really helpful to have this reach visible status by 17 May, for discussion at meetup in Barcelona on Certification and Audit on 17-18 May.

### CODE REVIEW

**Review by rphair — APPROVED** [review]
Looks good to me, especially since already extensively reviewed.  I do remember when this doc was posted on the forum & I think there were no comments there because it already looks complete.

Ordinarily I would recommend adding the forum link (https://forum.cardano.org/t/cip-proposal-cardano-audit-best-practice-guidelines/100022) as a `Comments-URI` but no sense doing that in this case because there's been no discussion there.

I'm not directly involved in the process that assigns CIP numbers but wanted to make sure this had 1 of the 2 required approving reviews: the next CIP meeting is on 10 May and probably @KtorZ and/or others could assign a number & push it through at that time, if not before.

Here is the link to that meeting if you would like to follow along or attend & speak, to assure it's approved & merged on time: https://www.crowdcast.io/e/cip-editors-meeting-44/register

File: CIP-Thompson-AuditBestPractice/CIP-XXX.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+Submitters shall provide specification and design documents that describe in a precise and unambiguous way the architecture, interfaces and requirements characterising the DApp. 
+
+The documentation shall identify the expected behaviour of the code, given without direct reference to the code itself. The description should also include high-level examples demonstrating use cases of the DApp. All assumptions about how the DApp will be used will be described. The documentation shall identify and document all the interfaces with other components and services.
+
+Submitters might also wish to explain mitigating actions that they have taken to protect against potential failures and attacks to the DApp.
+
+#### *On-Chain Specification* 
+
+The format of transactions accepted by the smart contracts should be specified using the template [here](https://docs.google.com/document/d/1j1s17Sdb7i8NzhK5GvgvcZqYbBtUzJB9Qpalmk8xElY).
```
Comment (KtorZ) [inline CIP-Thompson-AuditBestPractice/CIP-XXX.md:115]: May we make that document a markdown document attached to this particular CIP (you may put multiple files in the CIP folder).

**Review by KtorZ — COMMENTED** [review]
Did a first round of very "surface review". Few suggestions already however:

- Perhaps provide a concrete example of an audit request for some fake project, just to illustrate a bit the guidelines. Some of the wording in the document is a bit subjective (e.g. 'clearly specify') and would really depends on developers' own standards I reckon. Having some example which show where to set the bar could be beneficial. 

- Would it be worth it to perhaps have a list of auditors and their point of contact at the end of the CIP? We could encourage auditing companies to come and add themselves here for visibility and at the same time, make it 'low effort' for interested projects to find auditors. 

-  I like the on-chain specification contract interface idea which could in itself, be a separate CIP defining a common wording to describe the interface of on-chain scripts (think, OpenAPI for Cardano contraccts, cc @scarmuega). 

- Wild idea also, borderline in scope with this CIP but an idea nonetheless, would it makes sense to standardize the process by which auditors may confirm / advertise their audits? via for example, an on-chain transaction with metadata recording on-chain versions (e.g. script hashes) of the contract that they have audited, when, as well as the result of the audit. This could leverage the chain to provide proofs of audits at given point in time.

File: CIP-Thompson-AuditBestPractice/CIP-XXX.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+***Deployment***: once a DApp is developed it is deployed by submitting the relevant on-chain scripts onto the Cardano blockchain as transaction validator scripts. The scripts being deployed  might be the scripts that have been audited, or be instances of them in the case that the audited scripts are *parametric*.
+
+###  Tendering
+In order to provide a quote for audit, developers will need to supply
+* A specification of the DApp to be audited (more details below).
+* The scope of the audit.
+* An estimate of the scale of the audit work, e.g. the number of lines in the on-chain code to be audited, or the code itself, in its current state of development.
+
+### Submission
+In order to be audited, developers will need to supply the following documentation.
```
[the remark this answers is not in the document]
Comment (simonjohnthompson) [reply CIP-Thompson-AuditBestPractice/CIP-XXX.md:103]: Happy to do that.

File: CIP-Thompson-AuditBestPractice/CIP-XXX.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
+---
+CIP: \?
+Title: Cardano audit best practice guidelines
+Authors: Simon Thompson <simon.thompson@iohk.io>
+Comments-Summary: No comments yet.
+Comments-URI: https://github.com/cardano-foundation/CIPs/wiki/Comments:CIP-\?
+Status: Draft
+Type: Process
```
[the remark this answers is not in the document]
Comment (simonjohnthompson) [reply CIP-Thompson-AuditBestPractice/CIP-XXX.md:8]: Fine with me.

File: CIP-Thompson-AuditBestPractice/CIP-XXX.md (hunk 1 above)
Comment (simonjohnthompson) [reply CIP-Thompson-AuditBestPractice/CIP-XXX.md:115]: Certainly we could do that. We could also make it a separate CIP. I'll do the former for the moment.

**Comment by simonjohnthompson** [comment]
> Did a first round of very "surface review". Few suggestions already however:
> 
> * Perhaps provide a concrete example of an audit request for some fake project, just to illustrate a bit the guidelines. Some of the wording in the document is a bit subjective (e.g. 'clearly specify') and would really depends on developers' own standards I reckon. Having some example which show where to set the bar could be beneficial.

Happy to do that.

> * Would it be worth it to perhaps have a list of auditors and their point of contact at the end of the CIP? We could encourage auditing companies to come and add themselves here for visibility and at the same time, make it 'low effort' for interested projects to find auditors.

My only concern is about whether that confers some sort of approval on those listed. This may be something that legal would be unhappy with, as might those excluded from the list in some way.

> * I like the on-chain specification contract interface idea which could in itself, be a separate CIP defining a common wording to describe the interface of on-chain scripts (think, OpenAPI for Cardano contraccts, cc @scarmuega).

Yes.

> * Wild idea also, borderline in scope with this CIP but an idea nonetheless, would it makes sense to standardize the process by which auditors may confirm / advertise their audits? via for example, an on-chain transaction with metadata recording on-chain versions (e.g. script hashes) of the contract that they have audited, when, as well as the result of the audit. This could leverage the chain to provide proofs of audits at given point in time.

That's precisely what we're aiming to do. The delay in this has come from discussions about ways in which DApps can be  registered on Cardano, which leads into questions of identity. Once that's resolved the plan is to build on top of that.

**Comment by simonjohnthompson** [comment]
Hi folks - is there any update on the meeting on 10 May? If we were able to get a number for this, and make it visible, that would be great.

**Comment by KtorZ** [comment]
I'll be assigning a tentitative number today and minutes from the meeting will be available soon enough too.

In brief: we brought attention to this CIP in the call for DApp developers to have a look. We think it makes sense and we'd like to move it as "ready for review" for the next call, so giving a full 2 weeks for people to review and contribute to the discussion. In the meantime, we'll assign a number to it and unless major concerns are raised against (which I see as quite unlikely), it'll move to the next phase in the next bi-weekly meeting.

File: CIP-Thompson-AuditBestPractice/CIP-XXX.md (hunk 1 above)
Comment (simonjohnthompson) [reply CIP-Thompson-AuditBestPractice/CIP-XXX.md]: Done.

**Comment by KtorZ** [comment]
@simonjohnthompson in terms of editorial work, you may also:

- Put all files in a directory `CIP-0052`
- Have the main CIP file named `README.md` in that folder
- Name other files however you like
- Change the status to `proposed` as I assume there won't be any substantial addition to do

**Comment by simonjohnthompson** [comment]
It would be good if the proposal was visible through the CIP website, albeit as a proposal. Thanks.

**Comment by rphair** [comment]
FYI to readers of this old thread: applying category `Meta` isn't a confirmation of that category... it's a red flag to recategorise this, and any others to which a `Governance` category (in turn inclusive of "audit" technical procedures) may be applicable according to this forthcoming CIP (hopefully ready to be released soon):
* https://github.com/cardano-foundation/CIPs/issues/937

It will be less important after that time to change the label of this PR after that's done... though any editor would be welcome to do it any time after the forthcoming `Category: Governance` label is created. 😎

**Pull request merged**